content/post/2019-11-21-readmes.Rmarkdown

```diff
@@ -0,0 +1,416 @@
+
+```{r metapkgs}
+meta <- pkgsearch::cran_packages(pkglist)
+meta <- meta %>%
+  dplyr::mutate(URL = strsplit(URL, "\\,")) %>%
+  tidyr::unnest(URL) %>%
+  dplyr::filter(stringr::str_detect(URL, "github\\.com")) %>%
+  dplyr::mutate(URL = stringr::str_remove_all(URL, "\\(.*")) %>%
+  dplyr::mutate(URL = stringr::str_remove_all(URL, "\\#.*")) %>%
+  dplyr::mutate(URL = trimws(URL)) %>%
+  dplyr::select(Package, Title, Date, Version,
+                URL) %>%
+  dplyr::mutate(path = urltools::path(URL)) %>%
+  dplyr::mutate(path = stringr::str_remove(path, "\\/$")) %>%
+  tidyr::separate(path, sep = "\\/", into = c("owner", "repo"))
+str(meta)
+```
+
+At this point we have `r nrow(meta)` packages with `r length(unique(meta$URL))` unique GitHub repo URLs, pfiew.
+
+We'll then extract their [preferred README from GitHub V3 API](https://developer.github.com/v3/repos/contents/#get-the-readme). Some of them won't even have one so we'll lose them from the sample. 
+
+```{r, get-readmes}
+
+gh <- memoise::memoise(ratelimitr::limit_rate(gh::gh,
+                                              ratelimitr::rate(1, 1)))
+get_readme <- function(owner, repo){
+  readme <- try(gh("GET /repos/:owner/:repo/readme",
+                   owner = owner, repo = repo),
+                silent = TRUE)
+  if(inherits(readme, "try-error")){
+    return(NULL)
+  }
+  
+  lines <- suppressWarnings(
+    readLines(readme$download_url
+              )
+    )
+  
+  if (length(lines) == 1){
+  sub(readme$path, lines,
+      readme$download_url) -> link
+  } else {
+    link <- readme$download_url
+  }
+  
+  tibble::tibble(owner = owner,
+                 repo = repo,
+                 readme = list(suppressWarnings(readLines(link))))
+}
+
+readmes <- purrr::map2_df(.x = meta$owner, .y = meta$repo,
+                          .f = get_readme)
+```
+
+The `readmes` data.frame has `r nrow(readmes)` lines so we lost a few more packages. 
+
+### Assessing README size
+
+#### Number of lines
+
+A first metric we'll extract is the number of lines of the README. 
+
+```{r}
+count_lines <- function(readme_lines){
+  readme_lines %>%
+    purrr::discard(. == "") %>% # emtpy lines
+    purrr::discard(stringr::str_detect(., "\\<\\!\\-\\-")) %>% # html comments
+    length()
+}
+
+readmes <- dplyr::group_by(readmes, owner, repo) %>%
+  dplyr::mutate(lines_no = count_lines(readme[[1]])) %>%
+  dplyr::ungroup()
+
+```
+
+How long are usual READMEs? Their number of lines range from `r min(readmes$lines_no)` to `r max(readmes$lines_no)` with a median of `r median(readmes$lines_no)`.
+
+```{r nolinesplot, fig.cap="Dot plot of the number of lines in READMEs"}
+library("ggplot2")
+
+ggplot(readmes) +
+  geom_dotplot(aes(lines_no), binwidth = 5) +
+  xlab("No. of lines") +
+  scale_y_continuous(NULL, breaks = NULL) +
+  hrbrthemes::theme_ipsum(base_size = 16,
+                          axis_title_size = 16) +
+  ggtitle(glue::glue("Number of lines in a sample of {nrow(readmes)} READMEs"))
+```
```

Comment:
Is there something wrong with the y scale here? Is it better to bin them, and show a simple histogram instead?

Reply:
I don't understand?

Reply:
![nolinesplot-1](https://user-images.githubusercontent.com/660288/70052746-d982e500-15cb-11ea-991b-39ff2330cc58.png)

Much of this plot is empty. I was wondering if a histogram would show more information.